alloc profiler: get stacks for pool allocs via wrapper function

vilterp opened the pull request:
merges into https://github.com/JuliaLang/julia/pull/42768

Add a wrapper around jl_gc_pool_alloc to allow us to instrument it.

`managed_malloc` taken care of in https://github.com/vilterp/julia/pull/20

## Before

![image](https://user-images.githubusercontent.com/7341/150021875-02900969-264c-44fd-a943-ebbf90116194.png)

*: instrumented; red: currently missed

<details>
<summary>graphviz</summary>

```graphviz
digraph G {
  node [shape=box];

  generated_code [label="<generated code>"];

  new_array;
  jl_gc_alloc_string [label="jl_gc_alloc_string *"];
  jl_gc_alloc [label="jl_gc_alloc *"];
  jl_managed_malloc [label="jl_managed_malloc *"]; // https://github.com/vilterp/julia/pull/20
  jl_gc_pool_alloc;
  jl_gc_big_alloc;

  generated_code -> jl_gc_alloc_string;
  generated_code -> new_array;
  generated_code -> jl_gc_alloc;
  generated_code -> jl_gc_big_alloc [color=red];
  generated_code -> jl_gc_pool_alloc [color=red];

  jl_gc_alloc -> jl_gc_pool_alloc;
  jl_gc_alloc -> jl_gc_big_alloc;
  new_array -> jl_gc_alloc;
  new_array -> jl_managed_malloc;
  jl_gc_alloc_string -> jl_gc_pool_alloc;
  jl_gc_alloc_string -> jl_gc_big_alloc;
}
```
</details>

## After

![image](https://user-images.githubusercontent.com/7341/150022015-ce23d9f0-86cc-4ca7-aece-428b815ee9b0.png)

*: instrumented; red: currently missed

<details>
<summary>graphviz</summary>
```graphviz
digraph G {
  node [shape=box];

  generated_code [label="<generated code>"];

  new_array;
  jl_gc_alloc_string [label="jl_gc_alloc_string *"];
  jl_gc_alloc [label="jl_gc_alloc *"];
  jl_managed_malloc [label="jl_managed_malloc *"]; // https://github.com/vilterp/julia/pull/20
  jl_gc_pool_alloc;
  jl_gc_pool_alloc_outer [fillcolor=lightgreen style=filled label="jl_gc_pool_alloc_outer *"];
  jl_gc_big_alloc;

  generated_code -> jl_gc_alloc_string;
  generated_code -> new_array;
  generated_code -> jl_gc_alloc;
  generated_code -> jl_gc_big_alloc [color=red];
  generated_code -> jl_gc_pool_alloc_outer;
  jl_gc_pool_alloc_outer -> jl_gc_pool_alloc;

  jl_gc_alloc -> jl_gc_pool_alloc;
  jl_gc_alloc -> jl_gc_big_alloc;
  new_array -> jl_gc_alloc;
  new_array -> jl_managed_malloc;
  jl_gc_alloc_string -> jl_gc_pool_alloc;
  jl_gc_alloc_string -> jl_gc_big_alloc;
}
```
</details>

vchuravy committed:
Do not fail if fastpath for TLS can't be found

Prevents loading of `libjulia-codegen` into `opt`.

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

vchuravy committed:
Fix cpu-features.ll opt invocation

vchuravy committed:
Make notunroll test robust against loop-rotation

vchuravy committed:
Revert "CI (Buildkite): temporarily allow the `llvmpasses` job to fail (#43663)"

This reverts commit 85a6990a9c1d49dd5aeaffeb4b38f881dc120823.

vtjnash committed:
tests: fix type-error in computing duration (#43691)

Regression introduced by ca11c2fc24a (when tests fail or are interrupted)

adigitoleo committed:
doc: add reference to `rethrow()` in `throw` docstring, closes #43669 (#43674)

bkmgit committed:
doc: add Xoshiro example (#43697)

dkarrasch committed:
Fix sparse array `setindex(::Int, ::Vector)` (#43678)

tfiers committed:
Docstring of `|>`: add parentheses around `->` func (#43699)

vchuravy committed:
Merge pull request #43689 from JuliaLang/vc/llvmpasses

Fix llvmpasses test

vtjnash committed:
macOS: workaround a dyld/libunwind deadlock issue since 12.1

Apple reintroduced the old bug that we previously worked around in
fad04d39d592d8e0fcbfba439e8157f582bbc850 with a similar patch to this.

This is needed anywhere that we may attempt to stop threads.

Fixes #43578

vtjnash committed:
[Test] avoid scanning Base twice for ambiguities (#43694)

dnadlinger committed:
Fix Base.StackTraces.lookup(C_NULL - 1) on macOS 12 (#43612)

See comment in diff for explanation. This fixes test/stacktraces.jl
on aarch64 macOS 12, and according to an OpenJDK issue where they
ran into the same problem, https://git.openjdk.java.net/jdk/pull/6193,
probably also x86_64 macOS 12.

ericphanson committed:
[Test] make TestLogger respect `maxlog` (#43641)

DilumAluthgeBot committed:
🤖 Bump the Pkg stdlib from fff928aa to 32880c47 (#43693)

Co-authored-by: Dilum Aluthge <dilum@aluthge.com>

IanButterworth committed:
replace inefficient logic (#43708)

aviatesk committed:
inference: add some error checks for atomic `getfield` with boundscheck (#43651)

vtjnash committed:
reflection: use a real world for lookups (#43695)

aviatesk committed:
inference: add error/throwness checks for `setfield!` (#43603)

In a similar spirit to #43587, this commit introduces error check for
`setfield!`. We can bail out from inference if we can prove either of:
- the object is not mutable type
- the object is `Module` object
- the value being assigned is incompatible with the declared type of
  object field

This commit also adds the throwness check for `setfield!` (i.e. `setfield!_nothrow`).
This throwness check won't be used in the current native compilation
pipeline since `setfield!` call can't be eliminated even if we can prove
that it never throws. But this throwness check would be used by
EscapeAnalysis.jl integration and so I'd like to include it in Base.

aviatesk committed:
inference: introduce more error/throwness checks for array primitives (#43587)

Obvious errors are usually caught in higher places like `convert`, but
it's better to have error checks at each builtin level in order to enable
early bail out from errorneous code compilation (when somehow it does not
rely on common abstraction). These checks are also useful for third
consumers like JET.

ericphanson committed:
Throw `CapturedException`s from asyncmap to avoid dropping the stacktrace (#42105)

pchintalapudi committed:
Generate memset and memmove intrinsics from ccalls (#43580)

staticfloat committed:
Add `get_preferences()` to load all prefs in the current environment (#43684)

Previously, we had the machinery to load per-UUID, but some tooling
needs to be able to load all preferences (without loading
`Preferences.jl`), so it's better to support being able to load
everything from just within Base.

aviatesk committed:
inference: follow up #43603, better `getfield_tfunc` impl (#43713)

DilumAluthge committed:
`base/version_git.sh`: fix the default value for `Base.GIT_VERSION_INFO` (for e.g. when Git is not available) (#43717)

* `base/version_git.sh`: fix the default value for `Base.GIT_VERSION_INFO` (for e.g. when Git is not available)

* Use single-quotes to avoid the need to backslash-escape the double quotes

dnadlinger committed:
Fix build against LLVM 14 (Git main) (#43722)

4c45f292a0e688f7b32b1116f9feefc95f93f14d had reintroduced
a use of getNumArgOperands(), which was already slated for
removal in (at least) LLVM 12.

aviatesk committed:
reflection: remove `_methods_by_ftype` with array arguments (#43710)

dnadlinger committed:
Align ccall() jl_set_task_tid return type with C declaration [nfc]

This shouldn't matter on any platform with a sensible ABI; just
happened across this when tracking down some other issue.

vtjnash committed:
src: fix minor headers issues (#43687)

vtjnash committed:
[REPL] handle failure of edit call gracefully (#43690)

dnadlinger committed:
Avoid warning due to unused jl_set_gc_and_wait() on Darwin [nfc]

Also makes it clear to the reader that signals-mach.c uses
a different approach entirely (which almost tripped me up
the first time I was reading through this part of the codebase).

tkf committed:
Add `setcpuaffinity(cmd, cpus)` for setting CPU affinity of subprocesses (#42469)

Support running commands with cpumask.

vchuravy committed:
Merge pull request #43742 from dnadlinger/fix-set-task-tid-return

Align ccall() jl_set_task_tid return type with C declaration [nfc]

vchuravy committed:
Merge pull request #43743 from dnadlinger/no-set-gc-and-wait-on-darwin

Avoid warning due to unused jl_set_gc_and_wait() on Darwin [nfc]

timholy committed:
Add internal docs for dump.c (#43656)

These comments & renamings are intended to make it easier to
understand (de)serialization of `*.ji` files.

JeffBezanson committed:
add missing codegen fallback functions and fix test (#43706)

fixes #43705

mkitti committed:
Fix ref to "git_version.jl" to "version_git.jl", minor typo (#43734)

DilumAluthge committed:
`InteractiveUtils.versioninfo()`: print the value of the `LD_LIBRARY_PATH` or `DYLD_LIBRARY_PATH` environment variables if they are set (#43609)

aviatesk committed:
don't create empty handler list when calling `show_error_hints` (#43733)

pchintalapudi committed:
[Codegen] Remove global nulls variables (#43757)

Co-authored-by: Prem Chintalapudi <premc@csail.mit.edu>

aviatesk committed:
follow up #43565 (#43755)

WrRan committed:
Simpler and More Clear ``Common Operations'' (#43741)

dnadlinger committed:
codegen: Avoid some uses of PointerType::getElementType() (#43721)

getElementType() will eventually go away with the LLVM switch
to opaque pointers.

The assertion in gen_cfun_wrapper is still checked within
CreateAlignedLoad.

DilumAluthgeBot committed:
🤖 Bump the Pkg stdlib from 32880c47 to c3139663 (#43763)

Co-authored-by: Dilum Aluthge <dilum@aluthge.com>

IanButterworth committed:
Add do-block syntax for Semaphore acquire. Add Semaphore tests (#43730)

Co-authored-by: Takafumi Arakaki <takafumi.a@gmail.com>

simeonschaub committed:
add syntax for empty n-dimensional arrays (#41618)

KlausC committed:
all permission masks get type UInt16 (#43726)

fredrikekre committed:
When given - as the program name, read program from stdin (#43191)

DilumAluthge committed:
`versioninfo`: make sure not to print the same environment variable twice (#43753)

DilumAluthge committed:
Add some comments to `Make.inc` that explicitly define what we mean when we say "assert build of Julia" (#43732)

* Add some comments to `Make.inc` that explicitly define what we mean when we say "assert build of Julia"
* Add a brief note to the devdocs with instructions on how to build an "assert build" of Julia
* Recommend putting the variables into the `Make.user` file

pchintalapudi committed:
Remove various uses of jl_LLVMContext in codegen (#43749)

Remove many frivolous uses of jl_LLVMContext.

KristofferC committed:
run a temp cleanup after the precompilation step (#43761)

Fixes #43702

vilterp committed:
try wrapping pool alloc

vchuravy committed:
deps: Add missing checksums for Clang (#43769)

eschnett committed:
Add missing colour `:light_white` (#43744)

Julia currently defines 15 of the 16 "basic" ANSI colours, omitting "bright white" (see https://en.wikipedia.org/wiki/ANSI_escape_code). This adds "bright white", which Julia would call "light white".

vilterp committed:
update exported functions

vtjnash committed:
macOS: extend the workaround to cover the dyld/exc_server deadlock issue, since 12.1

Later, we should probably switch to using mach_exc_server generated from
`mig mach_exc.defs`.

DilumAluthge committed:
CI (Buildkite): set both the `RR_LOG` and `RR_UNDER_RR_LOG` environment variables (#43775)

pchintalapudi committed:
Add TBAA caching to jl_codectx_t (#43764)

Co-authored-by: Prem Chintalapudi <premc@csail.mit.edu>

vchuravy committed:
Remove spurious at_show from loading test (#43774)

pchintalapudi committed:
Remove explicit references to jl_data_layout (#43756)

Co-authored-by: Prem Chintalapudi <premc@csail.mit.edu>

dkarrasch committed:
Send AdjOrTrans(BlasMatrix) oop-products to BLAS (#43435)

vtjnash committed:
promote ranges to the largest of the the start, step, or length (as applicable) (#43059)

Be careful to use `oneunit` instead of `1`, so that arithmetic on
user-given types does not promote first to Int.

Fixes #35711
Fixes #10554

pchintalapudi committed:
Remove global basic integer types (#43752)


Co-authored-by: Prem Chintalapudi <premc@csail.mit.edu>

ven-k committed:
Add and use `isaccessiblefile` for `isfile_casesensitive` (#42922)

Fixes #42876

jakobnissen committed:
docs: mention `const` field attribute of mutable struct (#43531)

Keno committed:
Allow branch folding to look at type information

Folds things like:

```
%a = complicated()::Const(true)
goto if not %a
```

JeffBezanson committed:
fix typo in doc string for `display` (#43781)

pchintalapudi committed:
Remove global thunk attribute (#43782)

Keno committed:
Fix bug where Conditional wasn't being marked as ConstAPI

pchintalapudi committed:
Remove global basic pointer types (#43754)

Co-authored-by: Prem Chintalapudi <premc@csail.mit.edu>

vchuravy committed:
Handle freeze instruction in FindBaseValue (#43728)

vtjnash committed:
fix ranges PR conflict (#43790)

We cannot safely use reverse, so we do not anymore. That is now causing
conflict between #43059 and #29842. And this is likely clearer anyways.

Closes #43788

vtjnash committed:
Merge pull request #43701 from JuliaLang/jn/43578

workaround a dyld/libunwind deadlock issue, since macOS 12.1

eschnett committed:
Define show for CartesianIndices (#43780)

mkitti committed:
Fix #43004 by commenting out broken throw tests (#43739)

t-bltg committed:
Add trailing nl for `display(::TextDisplay, ...)` (#43787)

Fixes #43766.

pchintalapudi committed:
Track memops more accurately (#43778)

Keno committed:
Fix nothrow modeling of have_fma (#43785)

Allows functions that make use of fma to be considered ConstAPI
(so long has both have_fma branches return the same constant).

DilumAluthge committed:
`InteractiveUtils.versioninfo(; verbose = true)`: remove unnecessary blank line before "Memory" (#43810)

tomtuamnuq committed:
Add description of local kw in global scope (#42794)

nickrobinson251 committed:
Show only the "Test Passed" message when a test passes (#43795)

This restores the behavior of how `at-test` shows a test pass in Julia
v1.0 - v1.6, reverting the display changes introduced when
storing more info in the `Pass` type in #36879, closes #43794.

KristofferC committed:
move SparseArrays to the external repo https://github.com/JuliaLang/SparseArrays.jl (#43813)

dkarrasch committed:
Cleanup `copy_oftype`-like functions in factorizations (#43700)

Co-authored-by: Daan Huybrechs <daan.huybrechs@kuleuven.be>

dkarrasch committed:
Optimize `permutedims` on *diagonals (#43792)

Co-authored-by: Michael Abbott <32575566+mcabbott@users.noreply.github.com>

IanButterworth committed:
Profile: Faster data dict lookup (#43805)

Liozou committed:
Allow reinterpreting singleton types (#43500)

aviatesk committed:
Merge pull request #43797 from ianatol/ia/condfolding

Allow branch folding to look at type information, fix Conditional bug

vtjnash committed:
Revert "Profile: Faster data dict lookup (#43805)" (#43814)

N5N3 committed:
Try to fix `fma` on windows (#43530)

* enable fma on Windows

pchintalapudi committed:
Clean up more uses of jl_LLVMContext (#43791)

fonsp committed:
[REPL] More accurate `ends_with_semicolon` (#43374)

* More accurate `ends_with_semicolon`

Co-authored-by: Steven G. Johnson <stevenj@mit.edu>

danielmatz committed:
Add specialization of foreach for tuples (#31901)

vchuravy committed:
[LLVM] Remove LLVM 11 patches from source-tree (#43824)

devmotion committed:
Use `GlobalRef` of `Core.CodeInfo` in `@generated` (#43823)

Co-authored-by: Simeon Schaub <schaub@mit.edu>

vchuravy committed:
[LLVM] Add support for building experimental targets (#43822)

Liozou committed:
Fix getindex and setindex! on 0-dimensional reinterpretarray (#43819)

Keno committed:
Small test cleanup (#43831)

Deconflict some identifiers, make test run properly in Main module.

Keno committed:
Fix splatnew test (#43835)

The function being tested currently throws:
```
julia> _construct_structwithsplatnew()
ERROR: TypeError: in new, expected Int64, got a value of type String
Stacktrace:
 [1] StructWithSplatNew
   @ ./REPL[15]:3 [inlined]
 [2] _construct_structwithsplatnew()
   @ Main ./REPL[16]:1
 [3] top-level scope
   @ REPL[18]:1
```

As a result, compiler precision improvements can cause the test
to fail, which is not intended. Fix the fieldtype of the struct
to make sure it doesn't throw.

ViralBShah committed:
Move sparse matrix tests to the new SparseArrays.jl repo (#43832)

* Move sparse matrix tests to the new SparseArrays.jl repo

fredrikekre committed:
Update Documenter to latest version. (#43843)

aviatesk committed:
InteractiveUtils: add const annotations (#43846)

DilumAluthgeBot committed:
🤖 Bump the SparseArrays stdlib from 205b770 to 16dd9bd (#43848)

Co-authored-by: Dilum Aluthge <dilum@aluthge.com>

vilterp committed:
avoid segfault by zeroing out type tag

tkf committed:
Fix typos in runtime_intrinsics.c (#43850)

bkamins committed:
Add defalg methods to resolve potential dispatch ambiguities (#43426)

We could also consider adding `defalg(v::AbstractArray{<:Number}) = DEFAULT_UNSTABLE`,
but it is unlikely that someone will want to do `Union` of `<:Number`
and something other than `Missing` that still would support comparison.

Relevant for https://github.com/JuliaStrings/InlineStrings.jl/issues/21
(and other custom types that will want to add `defalg` support allowing
for `Union` with `Missing`).

maleadt committed:
Simplify loading code from stdin during precompilation. (#43853)

Keno committed:
Move typeassert effect-free modeling to the proper place (#43830)

This was open-coded inside inlining, but the effect-free modeling
code should be with the other builtins (in preparation of calling
it from more places).

NHDaly commented on src/gc.c:
```diff
@@ -1196,6 +1196,16 @@ static NOINLINE jl_taggedvalue_t *add_page(jl_gc_pool_t *p) JL_NOTSAFEPOINT
 }
 
+// called into by LLVM-generated code
+JL_DLLEXPORT jl_value_t *jl_gc_pool_alloc_outer(jl_ptls_t ptls, int pool_offset,
+                                          int osize)
```
My main comment is about trying to find a more informative name for this, like maybe:
- `jl_gc_pool_alloc_from_gen`
- `jl_gc_pool_alloc_from_llvm_gen`
- `jl_gc_pool_alloc_from_generated`...
- `jl_gc_pool_alloc_that_will_be_called_from_generated_code`... i dunno...

Blech naming is hard

NHDaly reviewed (approved):
Awesome. So, _**I**_ think this looks great, totally solves the problem, and makes sense! 👍 

I do think it'd be good to get a second opinion. Thanks for all the reviews of late, @vchuravy -- would you be able to have a look here too? The PR description explains very well what we did here. ❤️

vchuravy commented on src/gc.c:
```diff
@@ -1196,6 +1196,16 @@ static NOINLINE jl_taggedvalue_t *add_page(jl_gc_pool_t *p) JL_NOTSAFEPOINT
 }
 
+// called into by LLVM-generated code
+JL_DLLEXPORT jl_value_t *jl_gc_pool_alloc_outer(jl_ptls_t ptls, int pool_offset,
```
I would prefer keeping the same name here, and renaming the inner `internal` implementation, which then also shouldn't be exported (and ideally inlined into this function).

Maybe call that one then `jl_gc_pool_alloc_impl`

NHDaly replied:
We talked about inlining it too, but we weren't sure about how to inline it _only_ into this function..

If we mark that one inline, it'll end up inlining into the normal callers of it, though.. So the _C_ functions (indicated in Pete's graph, above) that calls `jl_gc_pool_alloc` would now call `jl_gc_pool_alloc_impl` instead, and then it would be inlined.

So it'd become inlined into `jl_gc_alloc_` and `jl_gc_alloc_string`. Maybe that's okay?

Or would you want us to create yet a third separate wrapper function that gets called by those? So we would have three functions:
- `jl_gc_pool_alloc_from_generated`: called from generated code
- `jl_gc_pool_alloc`: called from `jl_gc_alloc_` and `jl_gc_alloc_string`
- `jl_gc_pool_alloc_impl`: called from `jl_gc_pool_alloc`
? Or is that too weird? And the extra inlining is actually fine?

N5N3 committed:
Fix `extrema(x; dims)` for inputs with `NaN/missing` (#43604)

* Define `extrema` using `mapreduce`; support `init`

* Fix tests for SparseArrays

* API clean and export `extrema!`

* Re-implement `reducedim_init` for extrema

* Merge `master` to pull in https://github.com/JuliaLang/SparseArrays.jl/pull/63

* Mark `BigInt` tests as broken

Co-authored-by: Milan Bouchet-Valat <nalimilan@club.fr>
Co-authored-by: Simeon Schaub <simeondavidschaub99@gmail.com>
Co-authored-by: Takafumi Arakaki <aka.tkf@gmail.com>
Co-authored-by: Tim Holy <tim.holy@gmail.com>

maleadt committed:
Remove have_fma from the public API exports. (#43854)

martinholters committed:
Fix `tmerge` for (partially) const types with undef fields (#43812)

When `tmerge` is applied to `Const`/`PartialStruct` and a field is
`Const` in one of the types and `Union{}` (i.e. undef) in the other,
the resulting field type must not be `Const` (as the resursively called
`tmerge` produces).

Fixes #43784.

vilterp committed:
Allocation profiler (#42768)

## Overview

Record the type and stack of every allocation (or only at a given sample interval), and return as Julia objects.

Alternate approach to existing alloc profiler PR: https://github.com/JuliaLang/julia/pull/33467
Complementary to garbage profiler PR: https://github.com/JuliaLang/julia/pull/42658 (maybe there's some nice way to meld them)

This may be reinventing the wheel from #33467, but I'm not sure why that one needs stuff like LLVM passes. I mimicked some stuff from it, but this was my attempt to get something up and running. Could easily be missing stuff.

## Usage:

```julia
using Profile.Allocs
res = Allocs.@profile sample_rate=0.001 my_func()
prof = Allocs.fetch()
# do something with `prof`
```

See also: https://github.com/JuliaPerf/PProf.jl/pull/46 for support for visualizing these.

Co-authored-by: Nathan Daly <nhdaly@gmail.com>

vilterp commented:
Closed in favor of https://github.com/JuliaLang/julia/pull/43868 cuz I couldn't change the base to point at JuliaLang/julia
